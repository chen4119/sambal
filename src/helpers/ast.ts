import {
    factory,
    createPrinter,
    createSourceFile,
    SyntaxKind,
    ScriptKind,
    ScriptTarget,
    ListFormat,
    NewLineKind,
    Expression,
    ObjectLiteralExpression,
    ArrayLiteralExpression,
    PropertyAssignment,
    NewExpression,
    Identifier,
    StringLiteral,
    NumericLiteral,
    BooleanLiteral,
    Modifier,
    VariableStatement,
    NodeFlags,
    TypeReferenceNode,
    TypeNode,
    KeywordTypeNode,
    ArrayTypeNode
} from "typescript";
import fs from "fs";

export const EXPORT_MODIFIER = factory.createModifier(SyntaxKind.ExportKeyword);

export function makeValueLiteral(value: any): Expression {
    if (Array.isArray(value)) {
        return arrayToArrayLiteral(value);
    }
    switch (typeof(value)) {
        case "string":
            return makeStringLiteral(value);
        case "number":
            return makeNumericLiteral(value);
        case "boolean":
            return makeBooleanLiteral(value);
        case "object":
            return objectToObjectLiteral(value);
        case "undefined":
            return makeIdentifier("undefined");    
        default:
            throw new Error(`Unrecognized type: ${typeof(value)}`);
    }
}

export function objectToObjectLiteral(object: any): ObjectLiteralExpression {
    const properties = [];
    for (const prop of Object.keys(object)) {
        properties.push(makePropertyAssignment(prop, makeValueLiteral(object[prop])));
    }
    return makeObjectLiteral(properties);
}

export function arrayToArrayLiteral(array: any[]): ArrayLiteralExpression {
    const values = array.map((v) => makeValueLiteral(v));
    return makeArrayLiteral(values);
}

export function makeObjectLiteral(properties): ObjectLiteralExpression {
    return factory.createObjectLiteralExpression(properties, true);
}

export function makeArrayLiteral(values: Expression[]): ArrayLiteralExpression {
    return factory.createArrayLiteralExpression(values, true);
}

export function makePropertyAssignment(propName: string, initializer): PropertyAssignment{
    return factory.createPropertyAssignment(
        makeIdentifier(propName),
        initializer
    );
}

export function makeIdentifier(text: string): Identifier {
    return factory.createIdentifier(text);
}

export function makeStringLiteral(text: string): StringLiteral {
    return factory.createStringLiteral(text);
}

export function makeNumericLiteral(number: number): NumericLiteral {
    return factory.createNumericLiteral(String(number));
}

export function makeBooleanLiteral(bool: boolean): BooleanLiteral {
    return bool ? factory.createTrue() : factory.createFalse();
}

export function makeTypeReferenceNode(typeName: string, args: TypeNode[]): TypeReferenceNode {
    return factory.createTypeReferenceNode(
        makeIdentifier(typeName),
        args
    );
}

export function makeStringKeywordTypeNode(): KeywordTypeNode<SyntaxKind.StringKeyword> {
    return factory.createKeywordTypeNode(SyntaxKind.StringKeyword);
}

export function makeUnknownKeywordTypeNode(): KeywordTypeNode<SyntaxKind.UnknownKeyword> {
    return factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword);
}

export function makeArrayTypeNode(node: KeywordTypeNode | ArrayTypeNode): ArrayTypeNode {
    return factory.createArrayTypeNode(node);
}

export function makeVariableStatement(
        modifiers: Modifier[],
        variableName: string,
        variableInitializer: Expression,
        flag = NodeFlags.Const
    ): VariableStatement {
    return factory.createVariableStatement(
        modifiers,
        factory.createVariableDeclarationList(
            [factory.createVariableDeclaration(
                makeIdentifier(variableName),
                undefined,
                undefined,
                variableInitializer
            )], flag
        )
    );
}

export function makeCallExpression(functionName: string, args: Expression[]) {
    return factory.createCallExpression(
        makeIdentifier(functionName),
        undefined,
        args
    );
}

export function makeExpressionStatement(expr: Expression) {
    return factory.createExpressionStatement(expr);
}

export function makeEqualsBinaryExpression(leftExpr: Expression, rightExpr: Expression) {
    return factory.createBinaryExpression(
        leftExpr,
        factory.createToken(SyntaxKind.EqualsToken),
        rightExpr
    );
}

export function makePropertyAccess(expr: Expression, propName: string) {
    return factory.createPropertyAccessExpression(expr, makeIdentifier(propName));
}

export function makeEnum(modifiers: Modifier[], enumName: string, values: string[]) {
    return factory.createEnumDeclaration(
        undefined,
        modifiers,
        makeIdentifier(enumName),
        values.map((v) => factory.createEnumMember(makeIdentifier(v), undefined))
    );
}

export function makeNew(expression: Expression, args: Expression[]): NewExpression {
    return factory.createNewExpression(
        expression,
        undefined,
        args
    );
}

export function writeJavascript(statements, output) {
    const tsPrinter = createPrinter({
        newLine: NewLineKind.LineFeed
    });
    const sourceFile = createSourceFile(
        output,
        "",
        ScriptTarget.ES2015,
        false,
        ScriptKind.JS
    );
    const tsSource = tsPrinter.printList(
        ListFormat.MultiLine,
        factory.createNodeArray(statements),
        sourceFile
    );

    fs.writeFileSync(output, tsSource, "utf-8");
}