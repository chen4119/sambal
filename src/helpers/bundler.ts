import webpack, { Configuration } from "webpack";
import path from "path";
import nodeExternals from "webpack-node-externals";
import {
    CACHE_FOLDER,
    OnBundleChanged
} from "./constant";
import { log } from "./log";

const outputConfig = {
    library: {
        type: "umd"
    },
    globalObject: 'this'
};

const nodeJsConfig = {
    externalsPresets: { node: true },
    externals: [nodeExternals()]
};

const webpackConfig = {
    resolve: {
        extensions: [".js", ".jsx"]
    },
    module: {
        rules: [
            {
                test: /\.(js|jsx)$/,
                exclude: /node_modules/,
                use: {
                    loader: "babel-loader",
                    options: {
                        presets: ["@babel/preset-env", "@babel/preset-react"],
                        plugins: ["@babel/plugin-transform-runtime"]
                    }
                }
            },
            {
                test: /\.css$/,
                use: ['style-loader', 'css-loader']
            },
            {
                test: /\.(png|svg|jpg|jpeg|gif)$/i,
                type: 'asset/resource',
            },
            {
                // Apply rule for fonts files
                test: /\.(svg|woff|woff2|ttf|otf|eot)$/,
                type: "asset/inline"
            },
        ]
    }
};

export async function bundleSambalFile(entry: string, outputPath: string, isExcludeNode: boolean = true) {
    return new Promise<void>((resolve, reject) => {
        const config: Configuration = {
            ...webpackConfig,
            ...isExcludeNode ? nodeJsConfig : { externalsPresets: { node: true } },
            entry: entry,
            mode: "production",
            output: {
                ...outputConfig,
                filename: path.basename(entry),
                path: outputPath
            }
        };
        const compiler = webpack(config);
        compiler.run((err, stats) => {
            parseStats(err, stats, () => {
                if (!err) {
                    resolve();
                } else {
                    reject(err);
                }
            });
        });
    });
}

export function watchSambalFile(entry: string, onChange: OnBundleChanged) {
    const config: Configuration = {
        ...webpackConfig,
        ...nodeJsConfig,
        entry: entry,
        mode: "development",
        devtool: "eval-source-map",
        output: {
            ...outputConfig,
            filename: "[name].[contenthash].js",
            path: path.resolve(process.cwd(), `./${CACHE_FOLDER}/watch`)
        }
    };
    const compiler = webpack(config);

    return compiler.watch({
        aggregateTimeout: 300,
        poll: 1000
    }, (err, stats) => {
        parseStats(err, stats, onChange);
    });
}

export async function bundleBrowserPackage(bundleConfig: any, outputPath: string) {
    return new Promise<object>((resolve, reject) => {
        const config: Configuration = {
            ...webpackConfig,
            entry: bundleConfig.entry,
            mode: "production",
            output: {
                ...outputConfig,
                filename: "[name].[contenthash].js",
                path: outputPath
            }
        };
        const compiler = webpack(config);
        compiler.run((err, stats) => {
            parseStats(err, stats, (isError, entry) => {
                if (!err) {
                    resolve(entry);
                } else {
                    reject(err);
                }
            });
        });
    });
}

export function getDevServerBrowserCompiler(bundleConfig: any, listener) {
    const config: Configuration = {
        ...webpackConfig,
        entry: bundleConfig.entry,
        mode: "development",
        devtool: "eval-source-map",
        output: {
            ...outputConfig,
            filename: "[name].[contenthash].js"
        },
        plugins: [listener]
    };
    return webpack(config);
}

export function parseBundleFilename(stats) {
    const entries: any = {};
    for (const key of Object.keys(stats.entrypoints)) {
        const mainAsset = stats.entrypoints[key].assets[0];
        entries[key] = mainAsset.name;
    }
    return entries;
}

function parseStats(err, stats, onChange: OnBundleChanged) {
    const info = stats.toJson();
    const entry = parseBundleFilename(info);
    if (err) {
        log.error(err.stack || err.message);
        onChange(true, entry);
        return;
    }
    if (stats.hasErrors()) {
        // log.error(info.errors);
        for (const err of info.errors) {
            log.error(err.message);
        }
        onChange(true, entry);
        return;
    }
    if (stats.hasWarnings()) {
        // console.warn("%O", ...info.warnings);
        for (const warning of info.warnings) {
            log.warn(warning.message);
        }
    }
    onChange(false, entry);
}