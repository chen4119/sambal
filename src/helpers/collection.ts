import { JSONLD_ID } from "sambal-jsonld";
import {
    NULL_VALUE,
    EMPTY_VALUE
} from "./constant";
import {
    isJsDate
} from "./util";


// const MAX_INDEX_VALUE_CHAR = 600;

/*
export function mapJsonToPartitions(expandedJsonLd, groupBy) {
    const groupByValues = [];
    for (const groupByField of groupBy) {
        groupByValues.push(expandedJsonLd[groupByField]);
    }
    return getPartitionKeys(groupBy, groupByValues);
}

function getPartitionKeys(groupByKeys, groupByValues) {
    const partitionKeys = [];
    const indexes = groupByValues.map(value => Array.isArray(value) ? value.length : 0);
    let isDone = false;
    while (!isDone) {
        const partitionKey = [];
        for (let i = 0; i < groupByKeys.length; i++) {
            const value = groupByValues[i];
            if (Array.isArray(value)) {
                partitionKey.push(normalizeValue(value[--indexes[i]]));
            } else {
                partitionKey.push(normalizeValue(value));
            }
        }
        if (partitionKey.join("").length > MAX_INDEX_VALUE_CHAR) {
            throw new Error(`Partition key cannot excceed ${MAX_INDEX_VALUE_CHAR} chars`);
        }
        partitionKeys.push(partitionKey);
        isDone = indexes.filter(i => i > 0).length === 0; 
    }
    return partitionKeys;
}*/

function normalizeValue(value) {
    if (value === NULL_VALUE || value === EMPTY_VALUE) {
        throw new Error(`${value} is reserved`);
    } else if (typeof(value) === "undefined" || value === null) {
        return NULL_VALUE;
    } else if (typeof(value) === "string" && value.length === 0) {
        return EMPTY_VALUE;
    } else if (isJsDate(value)) {
        return value.getTime();
    } else if (typeof(value) === "object") {
        if (!value[JSONLD_ID]) {
            throw new Error("No complex index value allowed");
        }
        return value[JSONLD_ID];
    }
    return value;
}

export function getSortKey(jsonLd, propName: string) {
    return normalizeValue(jsonLd[propName]);
}

