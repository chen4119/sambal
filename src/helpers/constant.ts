
/*
export const WebElement = {
    SiteNavigation: "SiteNavigation",
    AdBlock: "AdBlock",
    Footer: "Footer",
    Header: "Header",
    SideBar: "SideBar",
    // Table: "Table"
}

export const Action = {
    Search: "Search",
    Review: "Review",
    Like: "Like",
    Listen: "Listen",
    View: "View",
    Watch: "Watch",
    Subscribe: "Subscribe",
    Comment: "Comment",
    Reply: "Reply",
    Share: "Share",
}*/

export const SORT_ASC = "asc";
export const SORT_DESC = "desc";

export type SortBy = {
    prop: string,
    order: "asc" | "desc"
}

export type Theme = {
    name: string,
    options: object
};

export type Collection = {
    "@id": string,
    "@type"?: string,
    src: string | string[],
    groupBy?: string[],
    sortBy?: SortBy
}

export type PageNode = {
    "@id": string,
    "@type": string,
    url: string,
    mainEntity: any,
    hasPart: any[],
    [key: string]: any
}

export type OnBundleChanged = (isError: boolean, entry: any) => void;

export interface IHtmlSerializer {
    toHtml(renderResult: unknown);
}

export const CACHE_FOLDER = ".sambal";
export const OUTPUT_FOLDER = "public";
export const SAMBAL_ENTRY_FILE = "sambal.entry.js";
export const SAMBAL_SITE_FILE = "sambal.site.js";
export const WEBSOCKET_ADDR = "ws://localhost:3001/";

// collection
export const NULL_VALUE = "__null__";
export const EMPTY_VALUE = "__empty__";
