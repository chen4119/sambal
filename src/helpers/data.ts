import axios, { AxiosResponse } from "axios";
import yaml from "js-yaml";
// import marked from "marked";
import glob from "glob";
import {
    getAbsFilePath,
    isValidRelativePath,
    readFileAsBuffer,
    readTextFile,
    frontMatter,
    safeParseJson,
    getFileExt
} from "./util";
import { isAbsUri } from "sambal-jsonld";

enum SUPPORTED_CONTENT_TYPE {
    yaml,
    markdown,
    json,
    image
}

export function searchFiles(query: string | string[]): string[] {
    if (Array.isArray(query)) {
        const matchSet = query.reduce((accumulator, value) => {
            searchFiles(value).forEach(m => accumulator.add(m));
            return accumulator;
        }, new Set<string>());
        return Array.from(matchSet);
    }
    if (isAbsUri(query)) {
        return [query];
    }
    const matches = glob.sync(query);
    return matches.filter(m => isValidRelativePath(m));
}

export function inferUrl(uri: string) {
    if (!uri || typeof(uri) !== "string") {
        return "";
    }
    if (uri.startsWith("/pages")) {
        let url = uri.substring(6);
        if (isSupportedFile(url)) {
            url = url.substring(0, url.lastIndexOf("."));
        }
        if (url === "/index") {
            return "/";
        }
        if (url.endsWith("/index")) {
            url = url.substring(0, url.length - 6);
        }
        return url;
    }
    return uri;
}

export async function loadLocalFile(src: string) {
    if (!isSupportedFile(src)) {
        throw new Error(`Unsupported file type: ${src}`);
    }

    const contentType = getLocalFileContentType(src);
    let content;
    if (contentType === SUPPORTED_CONTENT_TYPE.image) {
        content = await readFileAsBuffer(getAbsFilePath(src));
    } else {
        content = await readTextFile(getAbsFilePath(src));
    }
    return parseContent(content, contentType);
}

export async function loadRemoteFile(src: string) {
    const response = await axios.get(src);
    return parseContent(response.data, getAxiosResponseContentType(response));
}

const IMAGE_EXT_REGEX = /.+(.jpg|.jpeg|.gif|.png|.webp)$/i;
const SUPPORTED_FILE_EXT_REGEX = /.+(.yaml|.yml|.json|.md)$/i;

export function isImageFile(src: string) {
    if (!src) {
        return false;
    }
    return src.match(IMAGE_EXT_REGEX);
}

export function isSupportedFile(src: string) {
    if (!src) {
        return false;
    }
    return isAbsUri(src) || 
    isImageFile(src) ||
    src.match(SUPPORTED_FILE_EXT_REGEX);
}

function getAxiosResponseContentType(response: AxiosResponse<any>): SUPPORTED_CONTENT_TYPE {
    const contentType = response.headers["content-type"];
    if (contentType) {
        switch (contentType) {
            case "text/yaml":
            case "text/yml":
            case "application/yaml":
            case "application/yml":
                return SUPPORTED_CONTENT_TYPE.yaml;
            case "application/json":
            case "application/ld+json":
                return SUPPORTED_CONTENT_TYPE.json;
            case "text/markdown":
                return SUPPORTED_CONTENT_TYPE.markdown;
            case "image/jpeg":
            case "image/webp":
            case "image/gif":
            case "image/png":
                return SUPPORTED_CONTENT_TYPE.image;
        }
    }
    throw `Unsupported content type ${contentType}`;
}

function getLocalFileContentType(src: string): SUPPORTED_CONTENT_TYPE {
    const ext = getFileExt(src);
    switch (ext) {
        case "yaml":
        case "yml":
            return SUPPORTED_CONTENT_TYPE.yaml;
        case "json":
            return SUPPORTED_CONTENT_TYPE.json;
        case "md":
            return SUPPORTED_CONTENT_TYPE.markdown;
        case "jpg":
        case "jpeg":
        case "webp":
        case "gif":
        case "png":
            return SUPPORTED_CONTENT_TYPE.image;
        default:
            throw `Unsupported file type with extension ${ext}`;
    }
}

// TODO: ensure markdown is only for creative work!
function parseContent(content: any, contentType: SUPPORTED_CONTENT_TYPE) {
    try {
        let md;
        switch (contentType) {
            case SUPPORTED_CONTENT_TYPE.yaml:
                return yaml.load(content);
            case SUPPORTED_CONTENT_TYPE.json:
                return typeof(content) === "string" ? safeParseJson(content) : content;
            case SUPPORTED_CONTENT_TYPE.markdown:
                md = frontMatter(content);
                return {
                    ...md.frontMatter ? yaml.load(md.frontMatter) as object : {},
                    text: md.content,
                    encodingFormat: "text/markdown"
                }
            case SUPPORTED_CONTENT_TYPE.image:
                return content;
                // return Buffer.from(content, 'binary');
            default:
                throw `Unsupported content type ${contentType}.  Expecting yaml, json or markdown`;
        }
    } catch (e) {
        throw `Unable to parse content: ${e}`;
    }
}