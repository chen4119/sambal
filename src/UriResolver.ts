import { 
    IResolver,
    LOCALHOST,
    FS_PROTO,
    URI
} from "./helpers/constant";
import { URL } from "url";
import HttpResolver from "./resolvers/HttpResolver";
import FileSystemResolver from "./resolvers/FileSystemResolver";
import Media from "./Media";
import { getPathnameAndQuery, isObjectLiteral } from "./helpers/util";
import { 
    normalizeJsonLdId,
    isImageFile,
    isExternalSource
} from "./helpers/data";
import {
    JSONLD_CONTEXT,
    JSONLD_GRAPH,
    JSONLD_ID,
    JSONLD_TYPE,
    isSchemaType,
    isJsonLdRef
} from "sambal-jsonld";

type UriMatcher = {
    protocol?: string | string[],
    host?: string | string[],
    path?: string | string[]
}

const IMAGE_OBJECT = "imageobject";
const MAX_DEPTH = 100;

export default class UriResolver {
    private blankNodeIndex: number;
    private fsResolver: FileSystemResolver;
    private httpResolver: HttpResolver;
    private resolvers: {
        matcher: UriMatcher,
        resolver: IResolver
    }[];

    constructor(
        pages: string[],
        data: string[],
        private media: Media
    ) {
        this.blankNodeIndex = 1;
        this.fsResolver = new FileSystemResolver(pages, data);
        this.httpResolver = new HttpResolver();
        this.resolvers = [
            {
                matcher: {protocol: FS_PROTO},
                resolver: this.fsResolver
            },
            {
                matcher: {protocol: ["https", "http"]},
                resolver: this.httpResolver
            }
        ];
    }

    addResolver(matcher: UriMatcher, resolver: IResolver) {
        this.resolvers.unshift({
            matcher,
            resolver
        });
    }

    clearCache() {
        for (const instance of this.resolvers) {
            instance.resolver.clearCache();
        }
    }

    parseUri(uriStr: string) {
        const normalizedUriStr = normalizeJsonLdId(uriStr);
        let uri: URI;

        if (isExternalSource(normalizedUriStr)) {
            const url = new URL(uriStr);
            uri = {
                protocol: url.protocol,
                host: url.host,
                path: url.pathname,
                query: url.searchParams
            };
        } else {
            const pathAndQuery = getPathnameAndQuery(normalizedUriStr);
            uri = {
                protocol: FS_PROTO,
                host: LOCALHOST,
                path: pathAndQuery.pathname,
                query: pathAndQuery.query
            };
        }
        return uri;
    }

    async resolveUri(uriStr: string) {
        let uri: URI = this.parseUri(uriStr);
        let data;
        for (const instance of this.resolvers) {
            if (this.isMatch(instance.matcher, uri)) {
                data = await instance.resolver.resolveUri(uri);
                break;
            }
        }
        if (!data) {
            throw new Error(`No resolver found for uri ${uriStr}`);
        }
        if (this.fsResolver.isLocalImageFile(uri.path) || isImageFile(uriStr)) {
            data = await this.media.loadImageUrl(uri.path, data);
        } else if (isSchemaType(data, IMAGE_OBJECT, false)) {
            const image = await this.resolveUri(data.contentUrl);
            data = {
                ...data,
                ...await this.media.loadImageUrl(normalizeJsonLdId(data.contentUrl), image)
            };
        }
        return data;
    }

    private isMatch(matcher: UriMatcher, uri: URI) {
        return this.matcher(matcher.protocol, uri.protocol) &&
            this.matcher(matcher.host, uri.host) &&
            this.matcher(matcher.path, uri.path);
    }

    private matcher(expect: string | string[] | undefined, actual: string) {
        if (!expect) {
            return true;
        }
        return Array.isArray(expect) ? expect.indexOf(actual) >= 0 : expect === actual;
    }

    async hydrateUri(uri: string) {
        let jsonld = await this.resolveUri(uri);
        this.ensureJsonLd(jsonld, normalizeJsonLdId(uri));
        jsonld = await this.hydrate(jsonld);
        return jsonld;
    }

    // all data need @id and @type
    private async ensureJsonLd(jsonld: any, impliedId?: string) {
        if (Array.isArray(jsonld)) {
            return;
        }
        if (!jsonld[JSONLD_ID]) {
            jsonld[JSONLD_ID] = impliedId ? impliedId : `_:${this.blankNodeIndex++}`;
        }
    }

    async hydrate(
        target: unknown,
        level: number = 1,
        graph?: Map<string, unknown>
    ) {
        if (level > MAX_DEPTH) {
            throw new Error("Hydrating jsonld exceeded max depth level");
        }
        const nextLevel = level + 1;
        if (Array.isArray(target)) {
            const resolvedArr = [];
            for (const item of target) {
                resolvedArr.push(
                    await this.hydrate(item, nextLevel, graph)
                );
            }
            return resolvedArr;
        } else if (isJsonLdRef(target)) {
            let nextTarget;
            if (graph && graph.has(target[JSONLD_ID])) {
                nextTarget = graph.get(target[JSONLD_ID]);
            } else {
                nextTarget = await this.resolveUri(target[JSONLD_ID]);
                this.ensureJsonLd(nextTarget, normalizeJsonLdId(target[JSONLD_ID]));
            }
            return await this.hydrate(
                nextTarget,
                nextLevel,
                graph
            );
        } else if (isObjectLiteral(target)) {
            this.ensureJsonLd(target);
            await this.iterateObjectKeys(target, nextLevel, graph);
        }
        return target;
    }

    private async iterateObjectKeys(
        target: unknown,
        level: number,
        graph?: Map<string, unknown>
    ) {
        let objectGraph = null;
        for (const fieldName of Object.keys(target)) {
            if (fieldName !== JSONLD_ID && fieldName !== JSONLD_TYPE && fieldName !== JSONLD_CONTEXT) {
                const fieldValue = target[fieldName];
                if (fieldName === JSONLD_GRAPH) {
                    objectGraph = this.getGraphContext(fieldValue);
                }
                target[fieldName] = await this.hydrate(
                    fieldValue,
                    level,
                    objectGraph ? objectGraph : graph
                );
            }
        }
    }

    private getGraphContext(graph: any[]) {
        const context = new Map<string, unknown>();
        for (const item of graph) {
            context.set(item[JSONLD_ID], item);
        }
        return context;
    }
}