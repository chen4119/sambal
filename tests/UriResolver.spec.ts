import shelljs from "shelljs";
import UriResolver from "../src/UriResolver";
import { CACHE_FOLDER } from "../src/helpers/constant";
import { init } from "./setup";

describe("UriResolver", () => {
    let uriResolver: UriResolver;

    beforeEach(async () => {
        const classes = await init();
        uriResolver = classes.uriResolver
    });

    afterEach(async () => {
        shelljs.rm("-rf", CACHE_FOLDER);
    });

    it('get blog1', async () => {
        const result = await uriResolver.resolveUri("/blogs/blog1");
        expect(result).toMatchSnapshot();
    });

    it('hydrate blog1', async () => {
        const result = await uriResolver.hydrateUri("/blogs/blog1");
        expect(result).toMatchSnapshot();
    });

    it('get johnsmith', async () => {
        const result = await uriResolver.resolveUri("/john%20smith");
        expect(result).toMatchSnapshot();
    });

    it('hydrate johnsmith', async () => {
        const result = await uriResolver.hydrateUri("/john%20smith");
        expect(result).toMatchSnapshot();
    });

    it('get image2', async () => {
        const result = await uriResolver.resolveUri("/images/image2");
        expect(result).toMatchSnapshot();
        expect(shelljs.test('-f', `${CACHE_FOLDER}/images/image2.webp`)).toBeTruthy();
        expect(shelljs.test('-f', `${CACHE_FOLDER}/images/image2-50.webp`)).toBeTruthy();
    });

    it('get background', async () => {
        const result = await uriResolver.resolveUri("/background");
        expect(result).toMatchSnapshot();
    });

    it('hydrate background', async () => {
        const result = await uriResolver.hydrateUri("/background");
        expect(result).toMatchSnapshot();
    });

    it('hydrate header', async () => {
        const result = await uriResolver.hydrateUri("/header");
        expect(result).toMatchSnapshot();
    });

    it('get collections/tags', async () => {
        const collection = await uriResolver.resolveUri("/collections/tags?tag=java%20script");
        expect(collection["@id"]).toBe("/collections/tags/_part/tag=java%20script");
        expect(collection.itemListElement.length).toBe(2);
    });

    it('get collections/tags as nav', async () => {
        const collection = await uriResolver.resolveUri("/collections/tags?tag=java%20script&output=sitenav");
        expect(collection[0].url).toBe("/blogs/blog1");
        expect(collection[1].url).toBe("/blogs/blog2");
    });

    it('get collections/images', async () => {
        const collection = await uriResolver.resolveUri("/collections/images");
        expect(collection.itemListElement[0].contentUrl).toBe("/images/image2.webp");
    });

    describe("Custom resolver", () => {
        const exampleBlog = {
            "@type": "BlogPost",
            "text": "hello world"
        };

        const exampleResolver = {
            resolveUri: async (uri) => {
                return exampleBlog;
            }
        }

        beforeEach(() => {
            uriResolver.addResolver({host: "example.com"}, {
                resolveUri: exampleResolver.resolveUri,
                clearCache: () => {}
            });
        });

        it('get example.com/blog1', async () => {
            const result = await uriResolver.resolveUri("https://example.com/blog1");
            expect(result).toStrictEqual(exampleBlog);
        });
    });
});