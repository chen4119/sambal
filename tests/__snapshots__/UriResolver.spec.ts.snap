// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UriResolver get background 1`] = `
Object {
  "@type": "ImageObject",
  "author": Object {
    "@id": "/data/john smith.yml",
  },
  "contentSize": "399 kB",
  "contentUrl": "/data/images/image2.jpg",
  "encodingFormat": "image/jpeg",
  "height": 853,
  "name": "My background",
  "url": "https://example.com/data/images/image2.jpg",
  "width": 1280,
}
`;

exports[`UriResolver get blog1 1`] = `
Object {
  "@type": "BlogPosting",
  "author": Object {
    "@id": "/data/john smith.yml",
  },
  "dateCreated": 2019-10-26T00:00:00.000Z,
  "dateModified": 2019-10-26T00:00:00.000Z,
  "encodingFormat": "text/markdown",
  "headline": "My first blog",
  "image": Object {
    "@id": "https://dummyimage.com/300",
  },
  "keywords": Array [
    "rxjs",
    "java script",
  ],
  "position": 1,
  "text": "

Hello world",
}
`;

exports[`UriResolver get collections/tags 1`] = `
Array [
  Object {
    "@type": "BlogPosting",
    "author": Object {
      "@type": "Person",
      "email": "smitty123@hotmail.com",
      "familyName": "Smith",
      "givenName": "John",
      "name": "John Smith",
    },
    "dateCreated": 2019-10-26T00:00:00.000Z,
    "dateModified": 2019-10-26T00:00:00.000Z,
    "encodingFormat": "text/markdown",
    "headline": "My first blog",
    "image": Object {
      "@type": "ImageObject",
      "contentSize": "1 kB",
      "contentUrl": "https://dummyimage.com/300",
      "encodingFormat": "image/png",
      "height": 300,
      "url": "https://dummyimage.com/300",
      "width": 300,
    },
    "keywords": Array [
      "rxjs",
      "java script",
    ],
    "position": 1,
    "text": "

Hello world",
    "url": "/blogs/blog1",
  },
  Object {
    "@type": "BlogPosting",
    "author": Object {
      "@type": "Person",
      "email": "smitty123@hotmail.com",
      "familyName": "Smith",
      "givenName": "John",
      "name": "John Smith",
    },
    "dateCreated": 2017-10-26T00:00:00.000Z,
    "dateModified": 2017-10-26T00:00:00.000Z,
    "encodingFormat": "text/markdown",
    "headline": "My second blog",
    "image": Object {
      "@type": "ImageObject",
      "contentSize": "52 kB",
      "contentUrl": "/data/images/image2.webp",
      "encodingFormat": "image/webp",
      "height": 333,
      "thumbnail": Array [
        Object {
          "@type": "ImageObject",
          "contentSize": "2 kB",
          "contentUrl": "/data/images/image2-100w.webp",
          "encodingFormat": "image/webp",
          "height": 66,
          "url": "https://example.com/data/images/image2-100w.webp",
          "width": 100,
        },
        Object {
          "@type": "ImageObject",
          "contentSize": "746 B",
          "contentUrl": "/data/images/image2-50w.webp",
          "encodingFormat": "image/webp",
          "height": 33,
          "url": "https://example.com/data/images/image2-50w.webp",
          "width": 50,
        },
      ],
      "url": "https://example.com/data/images/image2.webp",
      "width": 500,
    },
    "keywords": Array [
      "sambal",
      "java script",
    ],
    "position": 2,
    "text": "

Hello world2",
    "url": "/blogs/blog2",
  },
]
`;

exports[`UriResolver get image2 1`] = `
Object {
  "@type": "ImageObject",
  "contentSize": "399 kB",
  "contentUrl": "/data/images/image2.jpg",
  "encodingFormat": "image/jpeg",
  "height": 853,
  "url": "https://example.com/data/images/image2.jpg",
  "width": 1280,
}
`;

exports[`UriResolver get image2 yml 1`] = `
Object {
  "@type": "ImageObject",
  "contentSize": "18 kB",
  "contentUrl": "/data/images/image2.webp",
  "encodingFormat": "image/webp",
  "height": 200,
  "representativeOfPage": true,
  "thumbnail": Array [],
  "url": "https://example.com/data/images/image2.webp",
  "width": 300,
}
`;

exports[`UriResolver get johnsmith 1`] = `
Object {
  "@type": "Person",
  "email": "smitty123@hotmail.com",
  "familyName": "Smith",
  "givenName": "John",
  "name": "John Smith",
}
`;

exports[`UriResolver hydrate background 1`] = `
Object {
  "@type": "ImageObject",
  "author": Object {
    "@type": "Person",
    "email": "smitty123@hotmail.com",
    "familyName": "Smith",
    "givenName": "John",
    "name": "John Smith",
  },
  "contentSize": "399 kB",
  "contentUrl": "/data/images/image2.jpg",
  "encodingFormat": "image/jpeg",
  "height": 853,
  "name": "My background",
  "url": "https://example.com/data/images/image2.jpg",
  "width": 1280,
}
`;

exports[`UriResolver hydrate blog1 1`] = `
Object {
  "@type": "BlogPosting",
  "author": Object {
    "@type": "Person",
    "email": "smitty123@hotmail.com",
    "familyName": "Smith",
    "givenName": "John",
    "name": "John Smith",
  },
  "dateCreated": 2019-10-26T00:00:00.000Z,
  "dateModified": 2019-10-26T00:00:00.000Z,
  "encodingFormat": "text/markdown",
  "headline": "My first blog",
  "image": Object {
    "@type": "ImageObject",
    "contentSize": "1 kB",
    "contentUrl": "https://dummyimage.com/300",
    "encodingFormat": "image/png",
    "height": 300,
    "url": "https://dummyimage.com/300",
    "width": 300,
  },
  "keywords": Array [
    "rxjs",
    "java script",
  ],
  "position": 1,
  "text": "

Hello world",
}
`;

exports[`UriResolver hydrate header 1`] = `
Object {
  "@type": "WPHeader",
  "hasPart": Array [
    Object {
      "@type": "SiteNavigationElement",
      "name": "About",
      "url": "/about",
    },
    Object {
      "@type": "SiteNavigationElement",
      "name": "Docs",
      "url": "/get-started",
    },
    Array [
      Object {
        "@type": "SiteNavigationElement",
        "name": "My first blog",
        "url": "/blogs/blog1",
      },
      Object {
        "@type": "SiteNavigationElement",
        "name": "My second blog",
        "url": "/blogs/blog2",
      },
    ],
  ],
  "headline": "Sambal",
  "potentialAction": Array [
    Object {
      "@type": "DownloadAction",
      "target": "https://github.com/chen4119/sambal",
    },
  ],
}
`;

exports[`UriResolver hydrate johnsmith 1`] = `
Object {
  "@type": "Person",
  "email": "smitty123@hotmail.com",
  "familyName": "Smith",
  "givenName": "John",
  "name": "John Smith",
}
`;
